### PR TITLE
Add automated testing with Jenkins

### PR DESCRIPTION
**Pull request by mjaehn**
This PR involves the following changes:

- [x] Change the folder structure for the test cases in a way that all created directories are within the root directory (i.e., `<path_to_chain>/work`,  `<path_to_chain>/output`, etc.
- [x] Non-comprehensive unification of config files
- [x] Read compute account from file or use standard one
- [x] Use pre-installed version of icontools
- [x] Get input data from ftp server
- [x] Use miniconda instead of pip and create environment file
- [x] Add script to build ICON
- [x] Update README and convert to Markdown
- [x] Set-up Jenkins plan: https://jenkins-mch.cscs.ch/view/C2SM/job/processing-chain-weekly/

### CODE REVIEW

File: README.md
Comment (efmkoene) [inline README.md:28]: The phrase in brackets is identical to the line before.

File: README.md
Comment (efmkoene) [inline README.md:60]: `your_account_id` -> `username`

File: README.md
Comment (efmkoene) [inline README.md:77]: Maybe also here mention that it's better to run this on `$SCRATCH`?

File: README.md
Comment (efmkoene) [inline README.md:84]: Does Spack work without `module load cray-python` on Daint?

File: README.md
Comment (efmkoene) [inline README.md:99]: Question (for everyone): do we keep `icon-art-BRM-CDOic`? It probably shouldn't be part of the master repository in its current state (e.g., with hard-coded links to specific executables etc.)

File: README.md
Comment (efmkoene) [reply README.md:60]: On CSCS you could instead write:
`echo $USER > ~/.acct` and `echo $USER@cscs.ch > ~/.forward`.

Comment (efmkoene) [reply README.md:60]: Or do you mean the computing project ID (one returned after typing `groups`) for acct?

File: cases/cosmo-ghg-11km-test/config.py
Comment (efmkoene) [inline cases/cosmo-ghg-11km-test/config.py:153]: Is this command going to work when spack hasn't been initialized with `source /project/g110/spack/user/daint/spack/share/spack/setup-env.sh`? Or is spack somehow initialized when activating the Conda environment?

File: cases/icon-art-oem-test/config.py
Comment (efmkoene) [inline cases/icon-art-oem-test/config.py:8]: Is there a reason this case isn't part of the Jenkins test cases?

File: README.md
Comment (mjaehn) [reply README.md:59]: Yes, the project ID is meant. I modified the description to make it more clear. For testing purposes, it is more convenient to have a standard account. For your own cases, you can always hardcode your desired account into the config file.

**Comment by efmkoene** [comment]
The one problem I have with the code right now is that all inputs, outputs, and intermediate results, are placed in a subfolder of the processing chain. I don't have enough space in my home folder to run this repository in its current form from my home folder. I will only be able to clone and test this repository on $SCRATCH, but that won't be ideal for development or use!
From that perspective, it could be better if data from [get_data.sh](https://github.com/C2SM/processing-chain/blob/jenkins/get_data.sh) was downloaded to $SCRATCH, and all outputs were also produced on $SCRATCH.

File: README.md
Comment (mjaehn) [reply README.md:76]: It is highly recommended now to clone the processing chain on `$SCRATCH` (as you would do for model source code). I will add this to the README. Taking track of your own changes / cases etc. should be done within your own fork.

File: README.md
Comment (mjaehn) [reply README.md:98]: Agreed. Could you move it to a personal fork?

File: cases/cosmo-art-mother-test/config.py
Comment (mjaehn) [reply cases/cosmo-art-mother-test/config.py:28]: For Jenkins, everything has to be within the root directory of the chain. Also, this is actually good for people to learn, as they easily find the `work/` and `output/` data etc.
This goes hand in hand with the comment above: The processing chain has to be cloned to `$SCRATCH`.

**Comment by gredvis** [comment]
A more general comment:  For some applications it would be good if the processing chain could be loaded as a python module , for example to make it available within another python package such as the inverse modelling frameworks CTDAS and CIF. Would it be a big effort to convert it into a python package?

File: cases/cosmo-ghg-11km-test/config.py
Comment (mjaehn) [reply cases/cosmo-ghg-11km-test/config.py:153]: According to the README, spack is initialized after activating the environment and before running the tests. However, there is a more elegant solution with `spack load cosmo@empa-ghg%nvhpc`. Then the path to the executable will be set as environment variable, which means we can write `cosmo_bin = cosmo-ghg_gpu`. In the COSMO runscript, we just have to initialize the spack environment there. Something similar has been done with `icontools`.
I will add this change.

File: cases/icon-art-oem-test/config.py
Comment (mjaehn) [reply cases/icon-art-oem-test/config.py:8]: They require a custom version of icon(-art). If @micstein89 send me instructions, which version to use, I can integrate it into the Jenkins testing.

**Comment by mjaehn** [comment]
Thanks @efmkoene for the thorough review! It is always appreciated to get input from the user's perspective! I understand that moving from `$HOME` to `$SCRATCH` seems inconvenient at the beginning, but there are good reasons behind doing so.

> I don't have enough space in my home folder to run this repository in its current form from my home folder. I will only be able to clone and test this repository on `$SCRATCH`, but that won't be ideal for development or use!

Again, the processing chain should now be used on on `$SCRATCH`. Could you elaborate why this is a problem for development or use? The `$HOME` directory on Piz Daint is a place for personal (backed-up) data and smaller scripts. If you stick to a proper Git(Hub) workflow, there is no problem working on  `$SCRATCH`. 

> The one problem I have with the code right now is that all inputs, outputs, and intermediate results, are placed in a subfolder of the processing chain.

This is only true for the test cases. You can design your own cases differently, specifying the input/output/work directories as you like.

> From that perspective, it could be better if data from [get_data.sh](https://github.com/C2SM/processing-chain/blob/jenkins/get_data.sh) was downloaded to $SCRATCH, and all outputs were also produced on $SCRATCH.

With that, Jenkins testing would not be possible or require dirty hacks.

> A more general comment: For some applications it would be good if the processing chain could be loaded as a python module , for example to make it available within another python package such as the inverse modelling frameworks CTDAS and CIF. Would it be a big effort to convert it into a python package?

There is some effort behind. Would someone from Empa be willing to do that? If not, it could be submitted as a C2SM task.

File: README.md
Comment (efmkoene) [reply README.md:98]: @IvaKsEmpa

File: cases/cosmo-ghg-11km-test/config.py
Comment (efmkoene) [reply cases/cosmo-ghg-11km-test/config.py:153]: OK. I'd suggest you update the README to make initializing spack a more clear requirement for running the chain for COSMO jobs. Right now it seems only required to build COSMO and not otherwise.

**Comment by efmkoene** [comment]
Thanks @mjaehn! I can see how one can work purely on `$SCRATCH` and can work via git commit/push for a reproducible setup, it just seems inconvenient to me that (untouched) files will be removed every 30 days, thus requiring a frequent synchronization between `$SCRATCH` and Github to deal with the temporary nature of the storage. But I understand the reasoning, and I'll see if I can work like that 👍

File: cases/cosmo-ghg-11km-test/config.py
Comment (mjaehn) [reply cases/cosmo-ghg-11km-test/config.py:153]: Normally, we could hide the initialization of spack in the runscript directly. However, this is not possible with int2lm as no executable `int2lm` is created. Thus, we indeed need to have spack initialized when running COSMO jobs. I added a note to the README.

Comment (efmkoene) [reply cases/cosmo-ghg-11km-test/config.py:153]: Great! 👍

File: env/environment.yml
Comment (mjaehn) [reply env/environment.yml:5]: I am testing this currently. If the miniconda build works, can you test the `cams4int2cosmo.py` afterwards?

**Comment by jcanton** [comment]
> Thanks @mjaehn! I can see how one can work purely on `$SCRATCH` and can work via git commit/push for a reproducible setup, it just seems inconvenient to me that (untouched) files will be removed every 30 days, thus requiring a frequent synchronization between `$SCRATCH` and Github to deal with the temporary nature of the storage. But I understand the reasoning, and I'll see if I can work like that 👍

I actually second @efmkoene opinion: for me one of the main reasons for using the chain is automatic copying to/from scratch. Due to the volatile nature of the filesystem, I find it much safer to have the chain with namelists and input files in /project or /store where it doesn't risk being accidentally deleted because I forgot a push or a touch, then one of the main jobs of the chain is automagically copying the necessary input files to scratch and running things there, and taking care of copying back the output files to /project or /scratch for safekeeping and postprocessing.
Similarly to the input/namelist files, I don't want to risk the deletion of an entire run because I forgot a touch, and I don't think it's appropriate to put in a repo the entire output files of a run

File: env/environment.yml
Comment (efmkoene) [reply env/environment.yml:5]: I'll test it tomorrow 👍

File: cases/cosmo-ghg-11km-test/config.py
Comment (efmkoene) [inline cases/cosmo-ghg-11km-test/config.py:26]: This comment is actually about four lines below, but Github doesn't allow me to make a comment there. 
`'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/cray/nvidia/default/lib64`
This has no effect because the added folder doesn't exist (anymore?) on Daint. Anyhow, it's also not needed (anymore?). So you can remove that line. It's also present in the test-case `cosmo-7` where you could remove it.

Comment (mjaehn) [reply cases/cosmo-ghg-11km-test/config.py:26]: Yes, there are probably some leftovers from past ages. You can try to make those changes and push directly to the `jenkins` branch.

File: README.md
Comment (mjaehn) [reply README.md:112]: @IvaKsEmpa @efmkoene I think we can safely remove it as you will be able to restore this case via Git history.

File: cases/icon-art-oem-test/config.py
Comment (mjaehn) [reply cases/icon-art-oem-test/config.py:8]: Any updates here? If it requires more work, we can integrate it in a separate PR.

File: env/environment.yml
Comment (efmkoene) [reply env/environment.yml:5]: It works perfectly, so this fixes issue #7 also.

File: README.md
Comment (IvaKsEmpa) [reply README.md:112]: I have my own fork , you can remove it

**Comment by mjaehn** [comment]
@efmkoene Is there anything more for this PR? If not, I will merge into master once one of you approves.
I think an icon-art(-oem) case can be added separately.

**Review by efmkoene — APPROVED** [review]
For me it's good to go!

**Pull request merged**
